<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item">

<li><span><a href="#For category prediction" data-toc-modified-id="For category prediction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Category Prediction</a></span></li><ul class="toc-item">

<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Random Prediction</a></span></li>
    
<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Item-to-Item KNN</a></span></li>
    
<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Seesion KNN</a></span></li>
 
<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>VMKNN</a></span></li></ul>

<li><span><a href="#For category prediction" data-toc-modified-id="For category prediction-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Article prediction</a></span></li><ul class="toc-item">

<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Random Prediction</a></span></li>
    
<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Item-to-Item KNN</a></span></li>
    
<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Seesion KNN</a></span></li>
 
<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>VMKNN</a></span></li></ul>

</ul></div>

In [11]:
import pandas as pd
import numpy as np
import sys
sys.path.append("./baseline")
import evaluation
import baselines
import sknn

import vmsknn
import iknn

# Category prediction

In [6]:
!python ./data/preprocess.py --Path ./data/day_one_category_region_device.csv --Feature True --Try True

Full train set
	Events: 190097
	Sessions: 67348
	Items: 40
Test set
	Events: 66876
	Sessions: 21499
	Items: 37
Train set
	Events: 141847
	Sessions: 51187
	Items: 40
Validation set
	Events: 48250
	Sessions: 16161
	Items: 33


In [2]:
data = pd.read_csv("./data/day_one_train_full.txt", sep='\t', dtype={'ItemId':np.int64})
test = pd.read_csv("./data/day_one_test_full.txt", sep='\t', dtype={'ItemId':np.int64})

## Random Prediction

In [8]:
RP=baselines.RandomPred()
RP.fit(data)
res = evaluation.evaluate_sessions(RP,test,data)
print("The accuracy of Random Prediction:")
print('Recall@20: {}'.format(res[0]))
print('MRR@20: {}'.format(res[1]))

100%|██████████| 66876/66876 [01:13<00:00, 906.62it/s] 

The accuracy of Random Prediction:
Recall@20: 0.22659056350133328
MRR@20: 0.07115760519191956


## Item-to-item KNN

In [9]:
iknn=iknn.ItemKNN(data)
iknn.fit(data)
res = evaluation.evaluate_sessions(iknn,test,data)
print("The accuracy of item k-nearest neighbors:")
print('Recall@20: {}'.format(res[0]))
print('MRR@20: {}'.format(res[1]))

100%|██████████| 66876/66876 [01:47<00:00, 621.44it/s]

The accuracy of item k-nearest neighbors:
Recall@20: 0.6575577935958745
MRR@20: 0.3679784283416388


## Session KNN

In [3]:
sknn=sknn.SessionKNN(data)
sknn.fit(data)
res = evaluation.evaluate_sessions(sknn,test,data)
print("The accuracy of session k-nearest neighbor:")
print('Recall@20: {}'.format(res[0]))
print('MRR@20: {}'.format(res[1]))    

100%|██████████| 13375/13375 [07:08<00:00, 31.19it/s]

The accuracy of session k-nearest neighbor:
Recall@20: 0.8080563504292317
MRR@20: 0.47766780569444295


## Vector Multiplication Session KNN

In [4]:
vmsknn=vmsknn.VMSessionKNN(data)
vmsknn.fit(data)
res = evaluation.evaluate_sessions(vmsknn,test,data)
print("The accuracy of session k-nearest neighbor:")
print('Recall@20: {}'.format(res[0]))
print('MRR@20: {}'.format(res[1]))    

100%|██████████| 13375/13375 [04:55<00:00, 45.31it/s]

The accuracy of session k-nearest neighbor:
Recall@20: 0.9318732115342285
MRR@20: 0.5027459391563475


# Article Prediction

In [8]:
!python ./data/preprocess.py --Path ./data/day_one_article_region_device.csv --Feature True --Try True

Full train set
	Events: 187978
	Sessions: 66718
	Items: 385
Test set
	Events: 35574
	Sessions: 11746
	Items: 267
Train set
	Events: 140275
	Sessions: 50721
	Items: 357
Validation set
	Events: 18508
	Sessions: 6735
	Items: 237


In [9]:
data = pd.read_csv("./data/day_one_train_full.txt", sep='\t', dtype={'ItemId':np.int64})
test = pd.read_csv("./data/day_one_test_full.txt", sep='\t', dtype={'ItemId':np.int64})

## Random Prediction

In [10]:
RP=baselines.RandomPred()
RP.fit(data)
res = evaluation.evaluate_sessions(RP,test,data)
print("The accuracy of Random Prediction:")
print('Recall@20: {}'.format(res[0]))
print('MRR@20: {}'.format(res[1]))

100%|██████████| 35574/35574 [00:40<00:00, 887.69it/s]

The accuracy of Random Prediction:
Recall@20: 0.02379553466509988
MRR@20: 0.0077107331377517675


## Item-to-item KNN

In [5]:
iknn=iknn.ItemKNN(data)
iknn.fit(data)
res = evaluation.evaluate_sessions(iknn,test,data)
print("The accuracy of item k-nearest neighbors:")
print('Recall@20: {}'.format(res[0]))
print('MRR@20: {}'.format(res[1]))

100%|██████████| 35574/35574 [00:51<00:00, 692.57it/s]

The accuracy of item k-nearest neighbors:
Recall@20: 0.4842622125230821
MRR@20: 0.18552524907338508


## Session KNN

In [6]:
sknn=sknn.SessionKNN(data)
sknn.fit(data)
res = evaluation.evaluate_sessions(sknn,test,data)
print("The accuracy of session k-nearest neighbor:")
print('Recall@20: {}'.format(res[0]))
print('MRR@20: {}'.format(res[1]))    

100%|██████████| 35574/35574 [05:13<00:00, 113.31it/s]

The accuracy of session k-nearest neighbor:
Recall@20: 0.5523753567231828
MRR@20: 0.1930098410300281


## VMKNN

In [7]:
vmsknn=vmsknn.VMSessionKNN(data)
vmsknn.fit(data)
res = evaluation.evaluate_sessions(vmsknn,test,data)
print("The accuracy of session k-nearest neighbor:")
print('Recall@20: {}'.format(res[0]))
print('MRR@20: {}'.format(res[1]))    

100%|██████████| 35574/35574 [03:53<00:00, 152.13it/s]

The accuracy of session k-nearest neighbor:
Recall@20: 0.6000503609199261
MRR@20: 0.2088733476155819


# Results

+ Category Prediction
    + Random Prediction:    
        Recall@20: 0.2266    
        MRR@20: 0.0712
    + Item-to-item KNN  
        Recall@20: 0.6576   
        MRR@20: 0.3680   
    + Session KNN     
        Recall@20: 0.8081     
        MRR@20: 0.4777    
    + VMSKNN
        Recall@20: 0.9319    
        MRR@20: 0.5027   
+ Article Prediction
    + Random Prediction
        Recall@20: 0.0237     
        MRR@20: 0.0078
    + Item-to-item KNN
        Recall@20: 0.4843   
        MRR@20: 0.1855  
    + Session KNN
        Recall@20: 0.5524    
        MRR@20: 0.1930      
    + VMSKNN
        Recall@20: 0.6001    
        MRR@20: 0.2089   